### PR TITLE
[XLA:GPU] Simplifies All-Reduce if they can be simplified or are degenerated.

### PR DESCRIPTION
**Pull request by yliu120**
This PR adds back the AllReduceSimplifier pass to the GPU compiler as in some cases we see JAX programs will add some degenerated psum to the HLOs,

Code pointer in JAX:
1) https://cs.opensource.google/jax/jax/+/main:jax/experimental/shard_map.py;drc=875f44c63a6c0f082ef3b228559e86e1d0a398d1;l=1674
2) pbroadcast to psum conversions. 

The compiler should generally remove these unnecessary ARs so that it unblocks many optimizations like collective pipeliner. It also emits unncessary runtime thunks as well.

### CODE REVIEW

**Comment by golechwierowicz** [comment]
I don't know the context. When was this pass removed?

**Comment by yliu120** [comment]
> I don't know the context. When was this pass removed?

I don't know as well... I think it should be there. At least it doesn't hurt anything.

**Comment by yliu120** [comment]
I was just fixing this issue so I don't check about the historical context.

**Comment by golechwierowicz** [comment]
Can you paste in a simple runnable JAX program which demonstrates the speedup? What speedups do you observe on such example and on which hardware?

**Review by ezhulenev — APPROVED** [review]

**Comment by ezhulenev** [comment]
It breaks bunch of internal tests, will try to land it tomorrow

**Comment by yliu120** [comment]
> It breaks bunch of internal tests, will try to land it tomorrow

OK. Thanks. I cannot see what is broken.. I guess something internally are not updated with this change.

**Comment by ezhulenev** [comment]
It's the constructor that got deleted

**Comment by yliu120** [comment]
> It's the constructor that got deleted

Yes. I removed that constructor because it does nothing. The field there is dead code and just confusing.

File: xla/service/gpu/tests/gpu_compiler_e2e_test.cc
Comment (golechwierowicz) [inline xla/service/gpu/tests/gpu_compiler_e2e_test.cc:1]: There's already a test file which captures gpu e2e pipeline and with collective ops testing in mind.

Can you move the test there: https://github.com/openxla/xla/blob/b527ed56098b3b1a7df99102c7e7c09d5f51c237/xla/tests/collective_ops_e2e_test.cc?

Comment (golechwierowicz) [reply xla/service/gpu/tests/gpu_compiler_e2e_test.cc:1]: Synced internally with @ezhulenev. We will rewrite the test internally.

Comment (yliu120) [reply xla/service/gpu/tests/gpu_compiler_e2e_test.cc:1]: Thank you both.

**Review by golechwierowicz — APPROVED** [review]

**Comment by golechwierowicz** [comment]
This is in https://github.com/openxla/xla/commit/23a255e065405ca12cde2824c09bf1fc7e4bfb0c.

**Pull request closed without merging**